### ISSUE TITLE
Not good enough syntax for loops (explicit lock needed)

### ISSUE THREAD

**Issue by emil14** [nevalang/neva#752]
> In this issue we are talking about looped network topologies rather than HOCs like  `map/filter/reduce`

# Usecase

In "99 bottles" example we have to use explicit lock instead of deferred connection because that's the only way to to implement manual loop:

`next2Lines` node should compute when we send initial message `99` and then each time it sends result which is not `-1` - in this case result must be send back to `next2Lines` (this is where to loop is created).

# Problem

We can't send `[99, next2Lines] -> switch` because `99` works as infinite sender and thus we'll create conflict between 2 streams, instead of sending 99 just once.

Because of this we need to defer sending `99` until `:start` is sent, which happens exactly once.

```neva
:start -> { 99 -> ??? }
[???, next2Lines] -> switch
```

Despite having a feature (deferred connections) exactly for that, we can't rely on it here and have to use explicit locks:

```neva
def Main(start) (stop) {
    Next2Lines, Lock<int> // explicit lock for fan-in to next2Lines
    ---
    :start -> lock:sig
    99 -> lock:data
    [lock:data, next2Lines] -> switch {
        -1 -> :stop
        else -> next2Lines
    }
}
```

---

Might be related to #717 

# Meta

This is not the perfect use case because it was caused by #754 concurrency issues with stream processing, so better way to deal with it is firstly to solve that problem and then observe if we need something else. Maybe all our problems could be solved with streams

**Comment by emil14** [comment]
# HACK: solve at the level of `next2Lines` by adding 1 more port

```neva
def Main(start) (stop) {
    Next2Lines
    ---
    :start -> { 99 -> next2Lines:start }
    next2Lines -> switch {
        -1 -> :stop
        else -> next2Lines:next
    }
}

def next2Lines(start int, next int) (n int) {
    Dec<int>, PrintFirstLine, PrintSecondLine
    ---
    // printFirstLine and printSecondLine won't work in parallel
    // because they are in the loop at the level of Main
    [:start, :next] -> printFirstLine -> dec -> printSecondLine -> :n
}
```

I don't like this solution, it's clearly a hack. We basically moved complexity from one place to another and probably made this even less obvious.

**Comment by emil14** [comment]
# HACK2: Introduce extra component

```
def Main(start) (stop) {
    Send99, Next2Lines
    ---
    :start -> send99
    [send99, next2Lines] -> switch {
        -1 -> :stop
        else -> next2Lines:next
    }
}

def Send99(sig any) (n int) {
    :sig -> { 99 -> :n }
}

def Next2Lines(start, next int) (n int) {
    Dec<int>, PrintFirstLine, PrintSecondLine
    ---
    // printFirstLine and printSecondLine won't work in parallel
    // because they are in the loop at the level of Main
    [:start, :next] -> printFirstLine -> dec -> printSecondLine -> :n
}
```

I would still consider this a hack, but maybe less than modifying `Next2Lines`. Perhaps somewhat at pair with original solution with explicit lock

**Comment by emil14** [comment]
# `Once` component

A component that receives messages but only sends one of them, discarding all the rest

```neva
def Once<T>(data T) (res T)
```

Could be used like this

```neva
def Main(start) (stop) {
    Once, Next2Lines
    ---
    :start -> { 99 -> once }
    [once, next2Lines] -> switch {
        -1 -> :stop
        else -> next2Lines:next
    }
}
```

## Questions

Should it send only first message completely ignoring all the rest or it should remember the message, compare new messages with it and send new as soon as it different than previous ones? Maybe Go has something like this? Or some other languages?

---

Related to #87

**Comment by emil14** [comment]
Could this be solved with for/foreach/etc? Some HOC or statement. Related to #723 and #739

**Comment by emil14** [comment]
At the moment solved with `While` (see #723)

This HOC is basically what implements loop with `lock` inside, so you don't have to

```neva
def Main(start) (stop) {
	While<int>{Next2Lines}
	---
	:start -> { 99 -> while:from }
	0 -> while:to
	while:res -> :stop
	while:err -> panic
}
```
